### ISSUE TITLE
Add a new template variable for URL subdomains and the base name

### ISSUE THREAD

**Issue by mikkovedru** [gildas-lormeau/SingleFile#1279]
**Is your feature request related to a problem? Please describe.**
I want to have my files to be saved in such a way that the pages from the same domain are close to each other. Regardless if I choose to have the domain as part of the directory or as starting part of the name. Let's assume that I am using the separate directory approach and let's imagine that I want to save the following URLs and this is the following order I want to get based on main domain name:
- A-product.**c**ompany.com/
- X-product.**c**ompany.com/
- articles.**r**euters.com/
- interviews.**r**euters.com/
- news.**r**euters.com/
- en.**w**ikipedia.org/
- it.**w**ikipedia.org/
- pt.**w**ikipedia.org/
- zh.**w**ikipedia.org/

Nice and logical. But what is the best I can get using the `{url-hostname}`?
- **A**-product.company.com/
- **a**rticles.reuters.com/
- **e**n.wikipedia.org/
- **i**nterviews.reuters.com/
- **i**t.wikipedia.org/
- **n**ews.reuters.com/
- **p**t.wikipedia.org/
- **X**-product.company.com/
- **z**h.wikipedia.org/

And now imagine thousands of different directories. It's a mess. The pages from the same domain can be worlds apart.

**Describe the solution you'd like**
Would be nice to have template variables for base name and the subdomains. So I could use e.g. `{url-basename} ({url-subname})` template and get:
- **c**ompany.com (A-product)/
- **c**ompany.com (X-product)/
- **r**euters.com (articles)/
- **r**euters.com (interviews)/
- **r**euters.com (news)/
- **w**ikipedia.org (en)/
- **w**ikipedia.org (it)/
- **w**ikipedia.org (pt)/
- **w**ikipedia.org (zh)/

Beautiful!

**Comment by gildas-lormeau** [comment]
Thank you for the suggestion, I agree that would be a useful feature. However, I'm wondering how subdomains of subdomains should be handled (e.g. `sub2.sub1.example.com`). I have a similar issue with `{url-last-segment}` (and previous segments in the path). Also, if you save a page on https://example.com, you'll get as folder name `example.com ()` in your example, but this might be an acceptable compromise.

**Comment by mikkovedru** [comment]
> how subdomains of subdomains should be handled (e.g. `sub2.sub1.example.com`)

I think that the best way is to treat the non-basename part as one. So the {url-subname} of `sub2.sub1.example.com` is `sub2.sub1`. Now that I think of it, the only useful change might (or might not) be to write it in the plural - `{url-subnames}`.


> if you save a page on [example.com](https://example.com/), you'll get as folder name example.com () in your example, but this might be an acceptable compromise.

Generally, most people would use the new subname variable in two ways:
1. To use the subname as a subdirectory - `/{url-basename}/{url-subname}/`. But in this case nothing needs to be done. If {url-subname} is missing, then `//` will change into nothing (`/example.com//index.html` = `/example.com/index.html`).
2. To use the subname as part of the name - `{url-basename} ({url-subname})`. This would indeed lead to `/example.com ()/`, which personally for me, is still an acceptable improvement. But what I would suggest is to implement a template language feature that would not only solve this particular issue, but would raise the template possibilities to another level: 
## Suggestion: Template language feature: output *with surroundings* (sic!) only if the info exists
The syntax could for example be `{before{url-subname}after}`.

Examples:

<table>
<tr>
 <th>Syntax example
 <th>The result if the item (e.g. subname) exists
 <th>The result if the item (e.g. subname) is empty
<tr>
 <td>/{url-basename}/{url-subname}/
 <td>/example.com/subname/
 <td>/example.com/
<tr>
 <td>/{url-basename}{ ({url-subname})}/
 <td>/example.com (subname)/
 <td>/example.com/
<tr>
 <td>/{url-basename}{_{url-subname}_(sic!)}/
 <td>/example.com_subname_(sic!)/
 <td>/example.com/
</table>

**Comment by gildas-lormeau** [comment]
Thank you, I agree with adding `{url-subdomain-names}`, `{url-root-domain}` and maybe `{url-last-subdomain-name}`, e.g. respectively `"sub2.sub1"`, `"example.com"` and `"sub2"` for the URL `sub2.sub1.example.com`.

However, I'm not so keen on your proposal concerning the management of characters before and after the variable. I think that, at this stage, a real template engine should be used so that functions could be used to manipulate strings, for example.

**Comment by gildas-lormeau** [comment]
I'm doing some tests with peg.js to improve the template engine, the syntax of a `if` function in the template would look like this `if<{url-subdomain-names},({url-subdomain-names})>`

**Comment by gildas-lormeau** [comment]
It's implemented, the new variables and the functions will be available in the next version.

**Comment by mikkovedru** [comment]
It is truly a marvel to see such dedication and speed! This is a great job, Gildas!

Here are my thoughts/comments/questions:
1. Can those functions be stacked? If yet, then it should be mentioned in documentation, with the example of using `%trim` after `%substring`. If not, then I couldn't come up with scenarios in which trim functions would be needed. Especially because:
2. All the template variable values should be automatically trimmed from both sides. If people want to use separate spaces, then can define them in the template. But by default values should not have spaces on the sides. I don't know if it is already done. If not, then it definitely should.
3. `%substring` documentation. Is it possible to leave one of the parameters empty (`%substring<{page-title}|5|>` = from 5th character until the end?
4. Inconsistency. On one hand it says `you can limit the length of a dynamic value by adding [maxByteSize] or [maxCharSizech] just after the function name`, but on another `"%substring<{page-title}|0|20>" to return the first 20 characters of the page title`.
5. This new functionality would really shine only after the ability to apply different templates based on whatever rules (most importantly - based on the original URL/domain of the page). Currently, the problem is that web sites are completely different and the best template for one site would be absolutely horrendous for another. But this functionality would allow me to create my own optimal rules for different sites. And since Pareto principle 80/20 creating only several separate rules for the web sites I use most (to read news and to save them), that would cover the most of my needs. The rest can be renamed manually. Have you thought about it? Do you have any plans? Regex would obviously be the best, but personally for my needs a simple "This {url-hostname-root} should use this template" would be enough.

Great job!

**Comment by gildas-lormeau** [comment]
I confirm you can stack function calls, I mean you can write for example `%if-empty<%capitalize<%page-element-attribute<meta[property="og:title"]|content>>|No title>.{filename-extension}`.
I agree that variable values should be trimmed but only for the template of filename, however it must not be trimmed for the template of the infobar.
The examples in the documentation have been written by CoPilot. I agree that it needs some improvements.

The default template in SingleFile is the same for more than 10 years. Moreover, folders may not work (e.g. on mobile). Maybe it would be interesting to propose somehow some default templates and explain what they do.

The proposition (6) is already possible via the auto-settings rules, AFAIK.
